ntoskrnl/ex/time.c

```diff
@@ -30,6 +30,113 @@ ULONG ExpTimerResolutionCount = 0;
 
 /* FUNCTIONS ****************************************************************/
 
+/*++
+ * If successful, this function sets the following Global Variables:
+ * ExpTimeZoneInfo
+ * ExpTimeZoneBias
+ * ExpTimeZoneId
+ *
+ * This variable is modified and returned as a result of this function
+ * *TimeZoneId
+ *--*/
+static
+BOOLEAN
+ExpGetTimeZoneId(
+    _In_ PLARGE_INTEGER TimeNow,
+    _Out_ PULONG TimeZoneId)
+{
+    LARGE_INTEGER StandardTime;
+    LARGE_INTEGER DaylightTime;
+    NTSTATUS Status;
+    static BOOLEAN InitialRun = TRUE;
+
+    /* Read time zone information from the registry */
+    Status = RtlQueryTimeZoneInformation(&ExpTimeZoneInfo);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        return FALSE;
+    }
+
+    if (!InitialRun)
+    {
+        /* Adjust TimeNow for Time Zone Bias to use UTC for comparisons */
+        TimeNow->QuadPart -= (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
+    }
+    else
+    {
+        InitialRun = FALSE;
+    }
+
+    /* Get the default bias */
+    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
```

Comment:
It may be of interest to get this default bias first, then do the InitialRun check and adjust the TimeNow with that bias: something like this:
```suggestion
    /* Get the default bias */
    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;

    if (!InitialRun)
    {
        /* Adjust TimeNow for Time Zone Bias to use UTC for comparisons */
        TimeNow->QuadPart -= ExpTimeZoneBias.QuadPart;
    }
    else
    {
        InitialRun = FALSE;
    }
```

Reply:
By the way, note that the LARGE_INTEGER variable passed via pointer : TimeNow, is also modified. I'm not sure this is correct.
You may have to locally capture its value in order to be able to internally use/modify it inside this function, without affecting what the caller of this function sees once the function returns.
By "locally capturing" the value, I mean, doing something like this:
```c
LARGE_INTEGER LocalTimeNow = *TimeNow;
...
LocalTimeNow.QuadPart -= ExpTimeZoneBias.QuadPart;
...
// Use LocalTimeNow instead of TimeNow in the rest of the code in this function.
```

Reply:
Changes made as suggested. Thanks.